drivers/platform/surface/surface_aggregator_registry.c

```diff
@@ -68,12 +68,26 @@ static const struct software_node ssam_node_bat_sb3base = {
 	.parent = &ssam_node_hub_base,
 };
 
-/* Platform profile / performance-mode device. */
-static const struct software_node ssam_node_tmp_pprof = {
+/* Platform profile / performance-mode device without a fan. */
+static const struct software_node ssam_node_tmp_perf_profile = {
 	.name = "ssam:01:03:01:00:01",
 	.parent = &ssam_node_root,
 };
 
+/* Platform profile / performance-mode device with a fan, such that
+ * the fan controller profile can also be switched.
+ */
+static const struct property_entry ssam_node_perf_profile_has_fan[] = {
+	PROPERTY_ENTRY_BOOL("has_fan"),
+	{ }
+};
+
+static const struct software_node ssam_node_tmp_perf_profile_with_fan = {
+	.name = "ssam:01:03:01:00:01",
+	.parent = &ssam_node_root,
+	.properties = ssam_node_perf_profile_has_fan,
```

Comment:
Not sure if that's a good idea, but apparently it's possible to directly put the properties here instead of defining an external array. [`[HWMON_CHANNEL_INFO()`](https://elixir.bootlin.com/linux/v6.6.9/C/ident/HWMON_CHANNEL_INFO)`](https://elixir.bootlin.com/linux/v6.6.9/source/include/linux/hwmon.h#L428) does something like that. Would reduce the amount of long names by one.

Reply:
Hmm, I tried this:

```c
	.properties = (((struct property_entry[]) {
		PROPERTY_ENTRY_BOOL("has_fan"),
		{}
	})),
```
But that doesn't compile because the `PROPERTY_ENTRY_BOOL` return is not considered const?
```
surface_aggregator_registry.c:89:51: error: initializer element is not constant
   89 |         .properties = (((struct property_entry[]) {
      |                                                   ^
```

I'd I'm not sure if this would be better either. The current is clear and readable, though a bit verbose. I renamed it from `ssam_node_perf_profile_has_fan` to  `ssam_node_tmp_perf_profile_has_fan` to have the `tmp` prefix in there for consistency.

Reply:
Yeah, I was mostly thinking that we maybe could reduce a bit of verbosity and make it look a bit nicer, given that that property is only used once. It should work if you use
```c
	.properties = (const struct property_entry[]) {
		PROPERTY_ENTRY_BOOL("has_fan"),
		{}
	},
```
The `properties` field is const (and you don't need those two outer braces).

But I'm also not sure if upstream is okay with that, at least I haven't found any instance of this being done before.

Reply:
Copy pasting the proposed code also gives:
```
surface_aggregator_registry.c:83:55: error: initializer element is not constant
   83 |         .properties = (const struct property_entry[]) {
      |                                                       ^
```

I'm fine with keeping this as is, given that I'm not sure how to express this in a way the compiler likes :P

I've changed the name to be `ssam_node_tmp_perf_profile_has_fan` for consistency with having the `tmp` in there.

Reply:
The way I understand it, values cannot be const, only variables. So the const in this case is a NOP.

Reply:
Hmm, this works for me when building as an external module and when building from `v6.6-surface`. Now that I'm checking it properly, it also works without the `const`... no clue why it doesn't work for you. Anyway, we can keep it as external array.

Reply:
Well, it may be a compiler issue? 

I made a minimal (non-)working example. With gcc 12.3 (and down to  in [godbolt](https://godbolt.org/z/rncnWcjvK), it reproduces my error:
```
<source>:61:51: error: initializer element is not constant
   61 |     .properties = (const struct property_entry[]) {
      |                                                   ^
<source>:61:51: note: (near initialization for 'our_node_nonconst_const')
Compiler returned: 1
```

But with gcc 13.1 it compiles fine [godbolt](https://godbolt.org/z/afYcbxsTf).

I started pruning compilation flags, but even until [gcc 10.1](https://godbolt.org/z/nvMdh8Kah) it won't compile...

Not sure what is happening, but it could explain why we don't see such a construct anywhere else in the codebase? Could be that one of the flags causes this error? This is just the default flags I get when compiling on NixOS.

Reply:
Huh, yeah. No idea why it works on one but not the other. Maybe some C standard stuff? Thanks for checking that. Let's keep it like it is then. I'll try to do a proper review some time in the next days.

Reply:
Friendly ping, trying to finish up my tablet setup and also wrap up outstanding PRs around it :)

If you're happy with this I'll rebase it on Torvald's `master` branch and send it to the `platform-driver-x86` list.

Reply:
Sorry! Had planned to take it up this weekend but a conference deadline got extended from march 1 until march 15 and things are currently a bit stressful in my day-job. Just had another quick look and I think it's ready.

Reply:
All good! I know open source stuff is best effort :) Hope your conference submission is going well.

I'll aim to get this sent upstream this weekend. Thanks for the review.